File: src/lib/services/insights.ts
import {
  getRecentEvents,
  getKnowledgeList,
  deleteKnowledge as deleteKnowledgeCommand,
  getTodoList,
  deleteTodo as deleteTodoCommand,
  generateDiary as generateDiaryCommand,
  deleteDiary as deleteDiaryCommand,
  getPipelineStats,
  getDiaryList,
  scheduleTodo as scheduleTodoCommand,
  unscheduleTodo as unscheduleTodoCommand,
  getEventCountByDate as getEventCountByDateCommand,
  getKnowledgeCountByDate as getKnowledgeCountByDateCommand,
  toggleKnowledgeFavorite as toggleKnowledgeFavoriteCommand,
  createKnowledge as createKnowledgeCommand,
  updateKnowledge as updateKnowledgeCommand
} from '@/lib/client/apiClient'

export interface InsightEvent {
  id: string
  title: string
  description: string
  keywords: string[]
  timestamp?: string
  createdAt?: string
  screenshots: string[]
}

export interface InsightKnowledge {
  id: string
  title: string
  description: string
  keywords: string[]
  mergedFromIds?: string[]
  createdAt?: string
  type?: 'combined' | 'original'
  deleted?: boolean
  favorite?: boolean
}

export interface RecurrenceRule {
  type: 'daily' | 'weekly' | 'monthly' | 'none'
  interval?: number
}

export interface InsightTodo {
  id: string
  title: string
  description: string
  keywords: string[]
  mergedFromIds?: string[]
  createdAt?: string
  completed?: boolean
  deleted?: boolean
  type?: 'combined' | 'original'
  scheduledDate?: string // YYYY-MM-DD format for calendar scheduling
  scheduledTime?: string // HH:MM format for time scheduling
  scheduledEndTime?: string // HH:MM format for end time
  recurrenceRule?: RecurrenceRule // Recurrence configuration
}

export interface InsightDiary {
  id: string
  date: string
  content: string
  sourceActivityIds: string[]
  createdAt?: string
}

interface ApiResponse<T> {
  success?: boolean
  message?: string
  data?: T
}

const ensureSuccess = <T>(response: ApiResponse<T>): T => {
  if (!response?.success) {
    throw new Error(response?.message ?? 'Unknown backend error')
  }
  if (!response.data) {
    throw new Error('Backend returned empty data')
  }
  return response.data
}

const normalizeScreenshots = (value: unknown): string[] => {
  if (!Array.isArray(value)) {
    return []
  }
  const images: string[] = []
  for (const item of value) {
    if (typeof item !== 'string') continue
    const trimmed = item.trim()
    if (!trimmed) continue
    images.push(trimmed)
    if (images.length >= 6) break
  }
  return images
}

const getScheduledDate = (value: Record<string, unknown>): string | undefined => {
  const snakeCase = typeof value.scheduled_date === 'string' ? value.scheduled_date : undefined
  const camelCase = typeof value.scheduledDate === 'string' ? value.scheduledDate : undefined
  const raw = snakeCase || camelCase
  return raw && raw.trim() ? raw : undefined
}

const getScheduledTime = (value: Record<string, unknown>): string | undefined => {
  const snakeCase = typeof value.scheduled_time === 'string' ? value.scheduled_time : undefined
  const camelCase = typeof value.scheduledTime === 'string' ? value.scheduledTime : undefined
  const raw = snakeCase || camelCase
  return raw && raw.trim() ? raw : undefined
}

const getScheduledEndTime = (value: Record<string, unknown>): string | undefined => {
  const snakeCase = typeof value.scheduled_end_time === 'string' ? value.scheduled_end_time : undefined
  const camelCase = typeof value.scheduledEndTime === 'string' ? value.scheduledEndTime : undefined
  const raw = snakeCase || camelCase
  return raw && raw.trim() ? raw : undefined
}

export async function fetchRecentEvents(limit: number, offset = 0): Promise<InsightEvent[]> {
  const raw = await getRecentEvents({ limit, offset })
  const data = ensureSuccess<{ events?: any[] }>(raw)
  const events = Array.isArray(data.events) ? data.events : []
  return events.map((event) => ({
    id: String(event.id ?? ''),
    title: String(event.title ?? ''),
    description: String(event.description ?? ''),
    keywords: Array.isArray(event.keywords) ? event.keywords : [],
    timestamp: typeof event.timestamp === 'string' ? event.timestamp : undefined,
    createdAt: typeof event.created_at === 'string' ? event.created_at : undefined,
    screenshots: normalizeScreenshots(event.screenshots)
  }))
}

export async function fetchKnowledgeList(): Promise<InsightKnowledge[]> {
  const raw = await getKnowledgeList()
  const data = ensureSuccess<{ knowledge?: any[] }>(raw)
  const knowledge = Array.isArray(data.knowledge) ? data.knowledge : []
  return knowledge.map((item) => ({
    id: String(item.id ?? ''),
    title: String(item.title ?? ''),
    description: String(item.description ?? ''),
    keywords: Array.isArray(item.keywords) ? item.keywords : [],
    mergedFromIds: Array.isArray(item.merged_from_ids) ? item.merged_from_ids : [],
    createdAt: typeof item.created_at === 'string' ? item.created_at : undefined,
    type: item.type === 'combined' ? 'combined' : 'original',
    deleted: Boolean(item.deleted),
    favorite: Boolean(item.favorite)
  }))
}

export async function deleteKnowledge(id: string) {
  const raw = await deleteKnowledgeCommand({ id })
  if (!raw?.success) {
    throw new Error(String(raw?.message ?? 'Failed to delete knowledge'))
  }
}

export async function toggleKnowledgeFavorite(id: string): Promise<InsightKnowledge> {
  const raw = await toggleKnowledgeFavoriteCommand({ id })
  if (!raw?.success || !raw.data) {
    throw new Error(String(raw?.message ?? 'Failed to toggle knowledge favorite'))
  }
  return {
    id: String(raw.data.id ?? ''),
    title: String(raw.data.title ?? ''),
    description: String(raw.data.description ?? ''),
    keywords: Array.isArray(raw.data.keywords) ? raw.data.keywords : [],
    createdAt: typeof raw.data.createdAt === 'string' ? raw.data.createdAt : undefined,
    favorite: Boolean(raw.data.favorite),
    deleted: Boolean(raw.data.deleted)
  }
}

export async function createKnowledge(
  title: string,
  description: string,
  keywords: string[]
): Promise<InsightKnowledge> {
  const raw = await createKnowledgeCommand({ title, description, keywords })
  if (!raw?.success || !raw.data) {
    throw new Error(String(raw?.message ?? 'Failed to create knowledge'))
  }
  return {
    id: String(raw.data.id ?? ''),
    title: String(raw.data.title ?? ''),
    description: String(raw.data.description ?? ''),
    keywords: Array.isArray(raw.data.keywords) ? raw.data.keywords : [],
    createdAt: typeof raw.data.createdAt === 'string' ? raw.data.createdAt : undefined,
    favorite: Boolean(raw.data.favorite),
    deleted: Boolean(raw.data.deleted)
  }
}

export async function updateKnowledge(
  id: string,
  title: string,
  description: string,
  keywords: string[]
): Promise<InsightKnowledge> {
  const raw = await updateKnowledgeCommand({ id, title, description, keywords })
  if (!raw?.success || !raw.data) {
    throw new Error(String(raw?.message ?? 'Failed to update knowledge'))
  }
  return {
    id: String(raw.data.id ?? ''),
    title: String(raw.data.title ?? ''),
    description: String(raw.data.description ?? ''),
    keywords: Array.isArray(raw.data.keywords) ? raw.data.keywords : [],
    createdAt: typeof raw.data.createdAt === 'string' ? raw.data.createdAt : undefined,
    favorite: Boolean(raw.data.favorite),
    deleted: Boolean(raw.data.deleted)
  }
}

export async function fetchTodoList(includeCompleted = false): Promise<InsightTodo[]> {
  const raw = await getTodoList({ includeCompleted })
  const data = ensureSuccess<{ todos?: any[] }>(raw)
  const todos = Array.isArray(data.todos) ? data.todos : []
  return todos.map((todo) => ({
    id: String(todo.id ?? ''),
    title: String(todo.title ?? ''),
    description: String(todo.description ?? ''),
    keywords: Array.isArray(todo.keywords) ? todo.keywords : [],
    mergedFromIds: Array.isArray(todo.merged_from_ids) ? todo.merged_from_ids : [],
    createdAt: typeof todo.created_at === 'string' ? todo.created_at : undefined,
    completed: Boolean(todo.completed),
    deleted: Boolean(todo.deleted),
    type: todo.type === 'combined' ? 'combined' : 'original',
    scheduledDate: getScheduledDate(todo),
    scheduledTime: getScheduledTime(todo),
    scheduledEndTime: getScheduledEndTime(todo),
    recurrenceRule: todo.recurrence_rule || undefined
  }))
}

export async function deleteTodo(id: string) {
  const raw = await deleteTodoCommand({ id })
  if (!raw?.success) {
    throw new Error(String(raw?.message ?? 'Failed to delete todo'))
  }
}

export async function scheduleTodo(
  todoId: string,
  scheduledDate: string,
  scheduledTime?: string,
  scheduledEndTime?: string,
  recurrenceRule?: RecurrenceRule
): Promise<InsightTodo> {
  const raw = await scheduleTodoCommand({
    todoId,
    scheduledDate,
    scheduledTime: scheduledTime && scheduledTime.trim() ? scheduledTime : undefined,
    scheduledEndTime: scheduledEndTime && scheduledEndTime.trim() ? scheduledEndTime : undefined,
    recurrenceRule
  } as any)
  const data = ensureSuccess<any>(raw)
  return {
    id: String(data.id ?? ''),
    title: String(data.title ?? ''),
    description: String(data.description ?? ''),
    keywords: Array.isArray(data.keywords) ? data.keywords : [],
    mergedFromIds: Array.isArray(data.merged_from_ids) ? data.merged_from_ids : [],
    createdAt: typeof data.created_at === 'string' ? data.created_at : undefined,
    completed: Boolean(data.completed),
    deleted: Boolean(data.deleted),
    type: data.type === 'combined' ? 'combined' : 'original',
    scheduledDate: getScheduledDate(data),
    scheduledTime: getScheduledTime(data),
    scheduledEndTime: getScheduledEndTime(data),
    recurrenceRule: data.recurrence_rule || undefined
  }
}

export async function unscheduleTodo(todoId: string): Promise<InsightTodo> {
  const raw = await unscheduleTodoCommand({ todoId })
  const data = ensureSuccess<any>(raw)
  return {
    id: String(data.id ?? ''),
    title: String(data.title ?? ''),
    description: String(data.description ?? ''),
    keywords: Array.isArray(data.keywords) ? data.keywords : [],
    mergedFromIds: Array.isArray(data.merged_from_ids) ? data.merged_from_ids : [],
    createdAt: typeof data.created_at === 'string' ? data.created_at : undefined,
    completed: Boolean(data.completed),
    deleted: Boolean(data.deleted),
    type: data.type === 'combined' ? 'combined' : 'original',
    scheduledDate: getScheduledDate(data),
    scheduledTime: getScheduledTime(data),
    scheduledEndTime: getScheduledEndTime(data),
    recurrenceRule: data.recurrence_rule || undefined
  }
}

export async function generateDiary(date: string): Promise<InsightDiary> {
  const raw = await generateDiaryCommand({ date })
  const data = ensureSuccess<any>(raw)
  return {
    id: String(data.id ?? ''),
    date: String(data.date ?? ''),
    content: String(data.content ?? ''),
    sourceActivityIds: Array.isArray(data.source_activity_ids) ? data.source_activity_ids : [],
    createdAt: typeof data.created_at === 'string' ? data.created_at : undefined
  }
}

export async function fetchDiaryList(limit: number): Promise<InsightDiary[]> {
  const raw = await getDiaryList({ limit })
  if (!raw?.success) {
    throw new Error(raw?.message ?? 'Unknown backend error')
  }
  const diaries = Array.isArray(raw.data?.diaries) ? raw.data.diaries : []
  return diaries.map((diary) => ({
    id: String(diary.id ?? ''),
    date: String(diary.date ?? ''),
    content: String(diary.content ?? ''),
    sourceActivityIds: Array.isArray(diary.sourceActivityIds) ? diary.sourceActivityIds : [],
    createdAt: typeof diary.createdAt === 'string' ? diary.createdAt : undefined
  }))
}

export async function deleteDiary(id: string) {
  const raw = await deleteDiaryCommand({ id })
  if (!raw?.success) {
    throw new Error(String(raw?.message ?? 'Failed to delete diary'))
  }
}

export async function fetchPipelineStats() {
  const raw = await getPipelineStats()
  return ensureSuccess(raw)
}

export async function fetchEventCountByDate(): Promise<Record<string, number>> {
  const raw = await getEventCountByDateCommand()
  const data = ensureSuccess<{ dateCountMap?: Record<string, number> }>(raw)
  return data.dateCountMap || {}
}

export async function fetchKnowledgeCountByDate(): Promise<Record<string, number>> {
  const raw = await getKnowledgeCountByDateCommand()
  const data = ensureSuccess<{ dateCountMap?: Record<string, number> }>(raw)
  return data.dateCountMap || {}
}

File: backend/handlers/insights.py
"""
Insights module command handlers (new architecture)
Insights module command handlers - handles events, knowledge, todos, diaries
"""

import uuid
from datetime import datetime
from typing import Any, Dict, List, Tuple

from core.coordinator import get_coordinator
from core.db import get_db
from core.logger import get_logger
from models.requests import (
    CreateKnowledgeRequest,
    DeleteItemRequest,
    GenerateDiaryRequest,
    GetDiaryListRequest,
    GetRecentEventsRequest,
    GetTodoListRequest,
    ScheduleTodoRequest,
    ToggleKnowledgeFavoriteRequest,
    UnscheduleTodoRequest,
    UpdateKnowledgeRequest,
)
from models.responses import (
    CreateKnowledgeResponse,
    DeleteDiaryResponse,
    DiaryData,
    DiaryListData,
    GenerateDiaryResponse,
    GetDiaryListResponse,
    KnowledgeData,
    ToggleKnowledgeFavoriteResponse,
    UpdateKnowledgeResponse,
)
from perception.image_manager import get_image_manager

from . import api_handler

logger = get_logger(__name__)


def get_pipeline():
    """Get new architecture processing pipeline instance"""
    coordinator = get_coordinator()
    coordinator.ensure_managers_initialized()
    pipeline = getattr(coordinator, "processing_pipeline", None)

    if pipeline is None:
        logger.error("Failed to get processing pipeline instance")
        raise RuntimeError("processing pipeline not available")

    return pipeline


def _get_data_access() -> Tuple[Any, Any]:
    """Return shared db + image manager instances"""
    db = get_db()
    image_manager = get_image_manager()
    return db, image_manager


async def _get_event_action_screenshot_hashes(db, event_id: str) -> List[str]:
    """Collect screenshot hashes for an event by looking up its source actions."""
    try:
        event = await db.events.get_by_id(event_id)
        if not event:
            return []

        action_ids = event.get("source_action_ids") or []
        if not action_ids:
            return []

        hashes: List[str] = []
        actions = await db.actions.get_by_ids(action_ids)
        for action in actions:
            hashes.extend(action.get("screenshots", []) or [])
        # Deduplicate while preserving order
        seen = set()
        deduped: List[str] = []
        for h in hashes:
            if h and h not in seen:
                seen.add(h)
                deduped.append(h)
        return deduped
    except Exception as exc:
        logger.error("Failed to load screenshot hashes for event %s: %s", event_id, exc)
        return []


async def _load_event_screenshots_base64(db, image_manager, event_id: str) -> List[str]:
    hashes = await _get_event_action_screenshot_hashes(db, event_id)

    screenshots: List[str] = []
    for img_hash in hashes:
        if not img_hash:
            continue
        data = image_manager.get_from_cache(img_hash)
        if not data:
            data = image_manager.load_thumbnail_base64(img_hash)
        if data:
            screenshots.append(data)

    return screenshots


# ============ Event Related Interfaces ============


@api_handler(
    body=GetRecentEventsRequest,
    method="POST",
    path="/insights/recent-events",
    tags=["insights"],
    summary="Get recent events",
    description="Get recent N event records (supports pagination)",
)
async def get_recent_events(body: GetRecentEventsRequest) -> Dict[str, Any]:
    """Get recent events

    @param body - Request parameters including limit and offset
    @returns Event list and metadata
    """
    try:
        db, image_manager = _get_data_access()
        limit = body.limit if hasattr(body, "limit") else 50
        offset = body.offset if hasattr(body, "offset") else 0

        events = await db.events.get_recent(limit, offset)
        for event in events:
            event["screenshots"] = await _load_event_screenshots_base64(
                db, image_manager, event["id"]
            )

        return {
            "success": True,
            "data": {
                "events": events,
                "count": len(events),
            },
            "timestamp": datetime.now().isoformat(),
        }

    except Exception as e:
        logger.error(f"Failed to get recent events: {e}", exc_info=True)
        return {
            "success": False,
            "message": f"Failed to get recent events: {str(e)}",
            "timestamp": datetime.now().isoformat(),
        }


# ============ Knowledge Related Interfaces ============


@api_handler(
    method="GET",
    path="/insights/knowledge",
    tags=["insights"],
    summary="Get knowledge list",
    description="Get all knowledge",
)
async def get_knowledge_list() -> Dict[str, Any]:
    """Get knowledge list

    @returns Knowledge list"""
    try:
        db, _ = _get_data_access()
        knowledge_list = await db.knowledge.get_list()

        return {
            "success": True,
            "data": {
                "knowledge": knowledge_list,
                "count": len(knowledge_list),
            },
            "timestamp": datetime.now().isoformat(),
        }

    except Exception as e:
        logger.error(f"Failed to get knowledge list: {e}", exc_info=True)
        return {
            "success": False,
            "message": f"Failed to get knowledge list: {str(e)}",
            "timestamp": datetime.now().isoformat(),
        }


@api_handler(
    body=DeleteItemRequest,
    method="POST",
    path="/insights/delete-knowledge",
    tags=["insights"],
    summary="Delete knowledge",
    description="Soft delete specified knowledge (including combined_knowledge)",
)
async def delete_knowledge(body: DeleteItemRequest) -> Dict[str, Any]:
    """Delete knowledge (soft delete)

    @param body - Contains knowledge ID to delete
    @returns Deletion result
    """
    try:
        db, _ = _get_data_access()
        await db.knowledge.delete(body.id)

        return {
            "success": True,
            "message": "Knowledge deleted",
            "timestamp": datetime.now().isoformat(),
        }

    except Exception as e:
        logger.error(f"Failed to delete knowledge: {e}", exc_info=True)
        return {
            "success": False,
            "message": f"Failed to delete knowledge: {str(e)}",
            "timestamp": datetime.now().isoformat(),
        }


@api_handler(
    body=ToggleKnowledgeFavoriteRequest,
    method="POST",
    path="/insights/toggle-knowledge-favorite",
    tags=["insights"],
    summary="Toggle knowledge favorite status",
    description="Toggle the favorite status of a knowledge item",
)
async def toggle_knowledge_favorite(body: ToggleKnowledgeFavoriteRequest) -> ToggleKnowledgeFavoriteResponse:
    """Toggle knowledge favorite status

    @param body - Contains knowledge ID
    @returns Updated knowledge data with new favorite status
    """
    try:
        db, _ = _get_data_access()
        new_favorite = await db.knowledge.toggle_favorite(body.id)

        if new_favorite is None:
            return ToggleKnowledgeFavoriteResponse(
                success=False,
                message="Knowledge not found",
                timestamp=datetime.now().isoformat(),
            )

        # Get updated knowledge data
        knowledge_list = await db.knowledge.get_list()
        knowledge_item = next((k for k in knowledge_list if k["id"] == body.id), None)

        if knowledge_item:
            knowledge_data = KnowledgeData(**knowledge_item)
            return ToggleKnowledgeFavoriteResponse(
                success=True,
                data=knowledge_data,
                message=f"Knowledge {'favorited' if new_favorite else 'unfavorited'}",
                timestamp=datetime.now().isoformat(),
            )
        else:
            return ToggleKnowledgeFavoriteResponse(
                success=False,
                message="Failed to retrieve updated knowledge",
                timestamp=datetime.now().isoformat(),
            )

    except Exception as e:
        logger.error(f"Failed to toggle knowledge favorite: {e}", exc_info=True)
        return ToggleKnowledgeFavoriteResponse(
            success=False,
            message=f"Failed to toggle knowledge favorite: {str(e)}",
            timestamp=datetime.now().isoformat(),
        )


@api_handler(
    body=CreateKnowledgeRequest,
    method="POST",
    path="/insights/create-knowledge",
    tags=["insights"],
    summary="Create knowledge manually",
    description="Create a new knowledge item manually",
)
async def create_knowledge(body: CreateKnowledgeRequest) -> CreateKnowledgeResponse:
    """Create knowledge manually

    @param body - Contains title, description, and keywords
    @returns Created knowledge data
    """
    try:
        db, _ = _get_data_access()

        # Generate unique ID
        knowledge_id = str(uuid.uuid4())
        created_at = datetime.now().isoformat()

        # Save knowledge
        await db.knowledge.save(
            knowledge_id=knowledge_id,
            title=body.title,
            description=body.description,
            keywords=body.keywords,
            created_at=created_at,
            source_action_id=None,  # Manual creation has no source action
            favorite=False,
        )

        # Return created knowledge
        knowledge_data = KnowledgeData(
            id=knowledge_id,
            title=body.title,
            description=body.description,
            keywords=body.keywords,
            created_at=created_at,
            source_action_id=None,
            favorite=False,
            deleted=False,
        )

        return CreateKnowledgeResponse(
            success=True,
            data=knowledge_data,
            message="Knowledge created successfully",
            timestamp=datetime.now().isoformat(),
        )

    except Exception as e:
        logger.error(f"Failed to create knowledge: {e}", exc_info=True)
        return CreateKnowledgeResponse(
            success=False,
            message=f"Failed to create knowledge: {str(e)}",
            timestamp=datetime.now().isoformat(),
        )


@api_handler(
    body=UpdateKnowledgeRequest,
    method="POST",
    path="/insights/update-knowledge",
    tags=["insights"],
    summary="Update knowledge",
    description="Update an existing knowledge item",
)
async def update_knowledge(body: UpdateKnowledgeRequest) -> UpdateKnowledgeResponse:
    """Update knowledge

    @param body - Contains knowledge ID, title, description, and keywords
    @returns Updated knowledge data
    """
    try:
        db, _ = _get_data_access()

        # Check if knowledge exists
        knowledge_list = await db.knowledge.get_list()
        knowledge_item = next((k for k in knowledge_list if k["id"] == body.id), None)

        if not knowledge_item:
            return UpdateKnowledgeResponse(
                success=False,
                message="Knowledge not found",
                timestamp=datetime.now().isoformat(),
            )

        # Update knowledge
        await db.knowledge.update(
            knowledge_id=body.id,
            title=body.title,
            description=body.description,
            keywords=body.keywords,
        )

        # Get updated knowledge
        knowledge_list = await db.knowledge.get_list()
        updated_knowledge = next((k for k in knowledge_list if k["id"] == body.id), None)

        if updated_knowledge:
            knowledge_data = KnowledgeData(**updated_knowledge)
            return UpdateKnowledgeResponse(
                success=True,
                data=knowledge_data,
                message="Knowledge updated successfully",
                timestamp=datetime.now().isoformat(),
            )
        else:
            return UpdateKnowledgeResponse(
                success=False,
                message="Failed to retrieve updated knowledge",
                timestamp=datetime.now().isoformat(),
            )

    except Exception as e:
        logger.error(f"Failed to update knowledge: {e}", exc_info=True)
        return UpdateKnowledgeResponse(
            success=False,
            message=f"Failed to update knowledge: {str(e)}",
            timestamp=datetime.now().isoformat(),
        )


# ============ Todo Related Interfaces ============


@api_handler(
    body=GetTodoListRequest,
    method="POST",
    path="/insights/todos",
    tags=["insights"],
    summary="Get todo list",
    description="Get all todos, optionally include completed",
)
async def get_todo_list(body: GetTodoListRequest) -> Dict[str, Any]:
    """Get todo list

    @param body - Request parameters, include include_completed
    @returns Todo list
    """
    try:
        db, _ = _get_data_access()
        include_completed = (
            body.include_completed if hasattr(body, "include_completed") else False
        )

        todo_list = await db.todos.get_list(include_completed)

        return {
            "success": True,
            "data": {"todos": todo_list},
            "timestamp": datetime.now().isoformat(),
        }

    except Exception as e:
        logger.error(f"Failed to get todo list: {e}", exc_info=True)
        return {
            "success": False,
            "message": f"Failed to get todo list: {str(e)}",
            "timestamp": datetime.now().isoformat(),
        }


@api_handler(
    body=DeleteItemRequest,
    method="POST",
    path="/insights/delete-todo",
    tags=["insights"],
    summary="Delete todo",
    description="Soft delete specified todo (including combined_todo)",
)
async def delete_todo(body: DeleteItemRequest) -> Dict[str, Any]:
    """Delete todo (soft delete)

    @param body - Contains todo ID to delete
    @returns Deletion result
    """
    try:
        db, _ = _get_data_access()
        await db.todos.delete(body.id)

        return {
            "success": True,
            "message": "Todo deleted",
            "timestamp": datetime.now().isoformat(),
        }

    except Exception as e:
        logger.error(f"Failed to delete todo: {e}", exc_info=True)
        return {
            "success": False,
            "message": f"Failed to delete todo: {str(e)}",
            "timestamp": datetime.now().isoformat(),
        }


@api_handler(
    body=ScheduleTodoRequest,
    method="POST",
    path="/insights/schedule-todo",
    tags=["insights"],
    summary="Schedule todo to a specific date",
    description="Set the scheduled_date for a todo",
)
async def schedule_todo(body: ScheduleTodoRequest) -> Dict[str, Any]:
    """Schedule todo to a specific date

    @param body - Contains todo ID, scheduled date, optional time, end time, and recurrence rule
    @returns Updated todo
    """
    try:
        db, _ = _get_data_access()
        updated_todo = await db.todos.schedule(
            body.todo_id,
            body.scheduled_date,
            body.scheduled_time,
            body.scheduled_end_time,
            body.recurrence_rule,
        )

        if not updated_todo:
            return {
                "success": False,
                "message": "Todo not found",
                "timestamp": datetime.now().isoformat(),
            }

        return {
            "success": True,
            "data": updated_todo,
            "message": "Todo scheduled successfully",
            "timestamp": datetime.now().isoformat(),
        }

    except Exception as e:
        logger.error(f"Failed to schedule todo: {e}", exc_info=True)
        return {
            "success": False,
            "message": f"Failed to schedule todo: {str(e)}",
            "timestamp": datetime.now().isoformat(),
        }


@api_handler(
    body=UnscheduleTodoRequest,
    method="POST",
    path="/insights/unschedule-todo",
    tags=["insights"],
    summary="Unschedule todo",
    description="Remove the scheduled_date from a todo",
)
async def unschedule_todo(body: UnscheduleTodoRequest) -> Dict[str, Any]:
    """Unschedule todo

    @param body - Contains todo ID
    @returns Updated todo
    """
    try:
        db, _ = _get_data_access()
        updated_todo = await db.todos.unschedule(body.todo_id)

        if not updated_todo:
            return {
                "success": False,
                "message": "Todo not found",
                "timestamp": datetime.now().isoformat(),
            }

        return {
            "success": True,
            "data": updated_todo,
            "message": "Todo unscheduled successfully",
            "timestamp": datetime.now().isoformat(),
        }

    except Exception as e:
        logger.error(f"Failed to unschedule todo: {e}", exc_info=True)
        return {
            "success": False,
            "message": f"Failed to unschedule todo: {str(e)}",
            "timestamp": datetime.now().isoformat(),
        }


# ============ Diary Related Interfaces ============


@api_handler(
    body=GenerateDiaryRequest,
    method="POST",
    path="/insights/generate-diary",
    tags=["insights"],
    summary="Generate diary",
    description="Generate diary for specified date based on all activities of that date",
)
async def generate_diary(body: GenerateDiaryRequest) -> GenerateDiaryResponse:
    """Generate diary

    @param body - Contains date (YYYY-MM-DD format)
    @returns Generated diary content
    """
    try:
        db, _ = _get_data_access()

        # Check if diary already exists
        diary = await db.diaries.get_by_date(body.date)

        if diary is not None:
            # Diary already exists, return it
            diary_data = DiaryData(**diary)
            return GenerateDiaryResponse(
                success=True,
                data=diary_data,
                timestamp=datetime.now().isoformat(),
            )

        # Diary doesn't exist, generate a new one
        # Get activities for the date
        activities = await db.activities.get_by_date(body.date, body.date)

        if not activities:
            return GenerateDiaryResponse(
                success=False,
                message=f"No activities found for date {body.date}",
                timestamp=datetime.now().isoformat(),
            )

        # Get DiaryAgent from coordinator and generate diary content
        coordinator = get_coordinator()
        coordinator.ensure_managers_initialized()
        diary_agent = coordinator.diary_agent

        if not diary_agent:
            return GenerateDiaryResponse(
                success=False,
                message="Diary agent not available",
                timestamp=datetime.now().isoformat(),
            )

        diary_content = await diary_agent.generate_diary(body.date, activities)

        if not diary_content:
            return GenerateDiaryResponse(
                success=False,
                message="Failed to generate diary content",
                timestamp=datetime.now().isoformat(),
            )

        # Extract activity IDs
        source_activity_ids = [activity["id"] for activity in activities]

        # Save diary to database
        diary_id = str(uuid.uuid4())
        await db.diaries.save(diary_id, body.date, diary_content, source_activity_ids)

        # Get the saved diary
        saved_diary = await db.diaries.get_by_date(body.date)

        if saved_diary:
            diary_data = DiaryData(**saved_diary)
            return GenerateDiaryResponse(
                success=True,
                data=diary_data,
                timestamp=datetime.now().isoformat(),
            )
        else:
            return GenerateDiaryResponse(
                success=False,
                message="Failed to retrieve saved diary",
                timestamp=datetime.now().isoformat(),
            )

    except Exception as e:
        logger.error(f"Failed to generate diary: {e}", exc_info=True)
        return GenerateDiaryResponse(
            success=False,
            message=f"Failed to generate diary: {str(e)}",
            timestamp=datetime.now().isoformat(),
        )


@api_handler(
    body=GetDiaryListRequest,
    method="POST",
    path="/insights/diaries",
    tags=["insights"],
    summary="Get diary list",
    description="Get recent diary records",
)
async def get_diary_list(body: GetDiaryListRequest) -> GetDiaryListResponse:
    """Get diary list"""
    try:
        db, _ = _get_data_access()
        diaries = await db.diaries.get_list(body.limit)

        # Convert diary dicts to DiaryData models
        diary_data_list = [DiaryData(**diary) for diary in diaries]

        return GetDiaryListResponse(
            success=True,
            data=DiaryListData(diaries=diary_data_list, count=len(diary_data_list)),
            timestamp=datetime.now().isoformat(),
        )

    except Exception as e:
        logger.error(f"Failed to get diary list: {e}", exc_info=True)
        return GetDiaryListResponse(
            success=False,
            message=f"Failed to get diary list: {str(e)}",
            timestamp=datetime.now().isoformat(),
        )


@api_handler(
    body=DeleteItemRequest,
    method="POST",
    path="/insights/delete-diary",
    tags=["insights"],
    summary="Delete diary",
    description="Delete specified diary",
)
async def delete_diary(body: DeleteItemRequest) -> DeleteDiaryResponse:
    """Delete diary

    @param body - Contains the diary ID to delete
    @returns Deletion result
    """
    try:
        db, _ = _get_data_access()
        await db.diaries.delete(body.id)

        return DeleteDiaryResponse(
            success=True,
            message="Diary deleted",
            timestamp=datetime.now().isoformat(),
        )

    except Exception as e:
        logger.error(f"Failed to delete diary: {e}", exc_info=True)
        return DeleteDiaryResponse(
            success=False,
            message=f"Failed to delete diary: {str(e)}",
            timestamp=datetime.now().isoformat(),
        )


# ============ Statistics Interface ============


@api_handler(
    method="GET",
    path="/insights/stats",
    tags=["insights"],
    summary="Get pipeline statistics",
    description="Get current pipeline runtime status and statistics data",
)
async def get_pipeline_stats() -> Dict[str, Any]:
    """Get pipeline statistics

    @returns pipeline runtime status and statistics data
    """
    try:
        pipeline = get_pipeline()
        stats = pipeline.get_stats()

        return {"success": True, "data": stats, "timestamp": datetime.now().isoformat()}

    except Exception as e:
        logger.error(f"Failed to get pipeline statistics: {e}", exc_info=True)
        return {
            "success": False,
            "message": f"Failed to get pipeline statistics: {str(e)}",
            "timestamp": datetime.now().isoformat(),
        }


@api_handler(
    method="GET",
    path="/insights/event-count-by-date",
    tags=["insights"],
    summary="Get event count by date",
    description="Get total event count for each date in database",
)
async def get_event_count_by_date() -> Dict[str, Any]:
    """Get event count grouped by date

    @returns Event count statistics by date
    """
    try:
        db, _ = _get_data_access()
        counts = await db.events.get_count_by_date()
        date_counts = [{"date": date, "count": count} for date, count in counts.items()]

        # Convert to map format: {"2025-01-15": 10, "2025-01-14": 5, ...}
        date_count_map = {item["date"]: item["count"] for item in date_counts}
        total_dates = len(date_count_map)
        total_events = sum(date_count_map.values())

        logger.debug(f"Event count by date: {total_dates} dates, {total_events} total events")

        return {
            "success": True,
            "data": {
                "dateCountMap": date_count_map,
                "totalDates": total_dates,
                "totalEvents": total_events
            },
            "timestamp": datetime.now().isoformat(),
        }
    except Exception as e:
        logger.error(f"Failed to get event count by date: {e}", exc_info=True)
        return {
            "success": False,
            "message": f"Failed to get event count by date: {str(e)}",
            "timestamp": datetime.now().isoformat(),
        }


@api_handler(
    method="GET",
    path="/insights/knowledge-count-by-date",
    tags=["insights"],
    summary="Get knowledge count by date",
    description="Get total knowledge count for each date in database",
)
async def get_knowledge_count_by_date() -> Dict[str, Any]:
    """Get knowledge count grouped by date

    @returns Knowledge count statistics by date
    """
    try:
        db, _ = _get_data_access()
        counts = await db.knowledge.get_count_by_date()
        date_counts = [{"date": date, "count": count} for date, count in counts.items()]

        # Convert to map format: {"2025-01-15": 10, "2025-01-14": 5, ...}
        date_count_map = {item["date"]: item["count"] for item in date_counts}
        total_dates = len(date_count_map)
        total_knowledge = sum(date_count_map.values())

        logger.debug(f"Knowledge count by date: {total_dates} dates, {total_knowledge} total knowledge")

        return {
            "success": True,
            "data": {
                "dateCountMap": date_count_map,
                "totalDates": total_dates,
                "totalKnowledge": total_knowledge
            },
            "timestamp": datetime.now().isoformat(),
        }
    except Exception as e:
        logger.error(f"Failed to get knowledge count by date: {e}", exc_info=True)
        return {
            "success": False,
            "message": f"Failed to get knowledge count by date: {str(e)}",
            "timestamp": datetime.now().isoformat(),
        }
